File: 2025/day2.1.py
def identify_invalid_ids(range_obj):
    """
    Find all numbers within the range where the second half equals the first half.
    Uses an optimized algorithm to jump through candidates instead of checking every number.
    """
    invalid_ids = []
    current = range_obj.start

    while current < range_obj.stop:
        num_str = str(current)
        num_digits = len(num_str)

        # If odd number of digits, jump to next number with even digits
        if num_digits % 2 == 1:
            current = 10 ** num_digits
            continue

        # Even number of digits - split into two halves
        half_len = num_digits // 2
        first_half = num_str[:half_len]
        second_half = num_str[half_len:]

        if first_half > second_half:
            # Replace second half with first half
            candidate = int(first_half + first_half)
        elif first_half == second_half:
            # Found an invalid ID
            invalid_ids.append(current)
            # Next candidate: increment first half by 1
            first_half_int = int(first_half) + 1
            # Check for overflow
            if len(str(first_half_int)) > half_len:
                # Overflow - jump to next power of 10
                candidate = 10 ** num_digits
            else:
                # No overflow - use incremented first half
                first_half_str = str(first_half_int).zfill(half_len)
                candidate = int(first_half_str + first_half_str)
        else:  # first_half < second_half
            # Increment first half by 1
            first_half_int = int(first_half) + 1
            # Check for overflow
            if len(str(first_half_int)) > half_len:
                # Overflow - jump to next power of 10
                candidate = 10 ** num_digits
            else:
                # No overflow - use incremented first half
                first_half_str = str(first_half_int).zfill(half_len)
                candidate = int(first_half_str + first_half_str)

        current = candidate

    return invalid_ids


def parse_ranges(line):
    """
    Parse a line of range pairs into a list of range objects.
    Example: "11-22,95-115" -> [range(11,23), range(95,116)]
    """
    ranges = []
    range_pairs = line.strip().split(',')

    for pair in range_pairs:
        first, last = pair.split('-')
        # Add 1 to last to make it inclusive
        ranges.append(range(int(first), int(last) + 1))

    return ranges


def main():
    # Read the file
    with open('day2.txt', 'r') as f:
        line = f.read().strip()

    # Parse ranges
    ranges = parse_ranges(line)

    # Find invalid IDs in each range and sum them up
    total = 0
    for range_obj in ranges:
        invalid_ids = identify_invalid_ids(range_obj)
        total += sum(invalid_ids)

    print(total)


if __name__ == '__main__':
    main()

File: 2025.go/day2.1.go
//go:build ignore

package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Range struct {
	first int
	last  int
}

func main() {
	data, err := os.ReadFile("../2025/day2.txt")
	if err != nil {
		log.Fatalf("Failed to read file: %v", err)
	}

	line := strings.TrimSpace(string(data))
	ranges := parseRanges(line)

	total := 0
	for _, r := range ranges {
		invalidIDs := identifyInvalidIDs(r)
		sum := 0
		for _, id := range invalidIDs {
			sum += id
		}
		total += sum
	}

	fmt.Println(total)
}

func parseRanges(line string) []Range {
	var ranges []Range
	pairs := strings.Split(line, ",")

	for _, pair := range pairs {
		parts := strings.Split(pair, "-")
		if len(parts) != 2 {
			log.Fatalf("Invalid range format: %s", pair)
		}

		first, err := strconv.Atoi(parts[0])
		if err != nil {
			log.Fatalf("Failed to parse first value: %s", parts[0])
		}

		last, err := strconv.Atoi(parts[1])
		if err != nil {
			log.Fatalf("Failed to parse last value: %s", parts[1])
		}

		// Add 1 to last to make it inclusive
		ranges = append(ranges, Range{first: first, last: last + 1})
	}

	return ranges
}

func identifyInvalidIDs(r Range) []int {
	var invalidIDs []int
	current := r.first

	for current < r.last {
		numDigits := countDigits(current)

		// If odd number of digits, jump to next power of 10
		if numDigits%2 == 1 {
			current = pow10(numDigits)
			continue
		}

		// Even number of digits - split in half
		halfDigits := numDigits / 2
		divisor := pow10(halfDigits)

		firstHalf := current / divisor
		secondHalf := current % divisor

		if firstHalf > secondHalf {
			// Replace second half with first half
			current = firstHalf*divisor + firstHalf
		} else if firstHalf == secondHalf {
			// Found an invalid ID
			invalidIDs = append(invalidIDs, current)
			// Move to next candidate: increase first half by 1
			firstHalf++
			if firstHalf >= divisor {
				// Overflow - jump to next power of 10
				current = pow10(numDigits)
			} else {
				current = firstHalf*divisor + firstHalf
			}
		} else {
			// firstHalf < secondHalf
			// Increase first half by 1 and replace second half
			firstHalf++
			if firstHalf >= divisor {
				// Overflow - jump to next power of 10
				current = pow10(numDigits)
			} else {
				current = firstHalf*divisor + firstHalf
			}
		}
	}

	return invalidIDs
}

func countDigits(n int) int {
	if n == 0 {
		return 1
	}
	count := 0
	for n > 0 {
		count++
		n /= 10
	}
	return count
}

func pow10(n int) int {
	result := 1
	for i := 0; i < n; i++ {
		result *= 10
	}
	return result
}
